soundwire: sdca: parse function type and controls

plbossart opened the pull request:
SDCA devices are based on independent 'Functions', which are exposed in platform firmware using the DisCo 2.1 specification.

This patch adds a mechanism to parse the functions and extract basic information (adr and function type).

Example log on a MeteorLake device (which already shows problems with ACPI information...)

[   13.069728] soundwire sdw:0:0:025d:0713:01: found invalid SDCA function type 4 ADR 1, skipped
[   13.069743] soundwire sdw:0:0:025d:0713:01: found SDCA function type 6 ADR 3
[   13.070166] soundwire sdw:0:1:025d:1316:01: found SDCA function type 1 ADR 4
[   13.070750] soundwire sdw:0:2:025d:1316:01: found SDCA function type 1 ADR 4

plbossart committed:
ASoC: sdca: parse function type and controls

SDCA devices are based on independent 'Functions', which are exposed
in platform firmware using the DisCo 2.1 specification.

This patch adds a mechanism to parse the functions and extract basic
information (adr and function type).

Example log on a MeteorLake device (which already shows problems with
ACPI information...)

[   13.069728] soundwire sdw:0:0:025d:0713:01: found invalid SDCA function type 4 ADR 1, skipped
[   13.069743] soundwire sdw:0:0:025d:0713:01: found SDCA function type 6 ADR 3
[   13.070166] soundwire sdw:0:1:025d:1316:01: found SDCA function type 1 ADR 4
[   13.070750] soundwire sdw:0:2:025d:1316:01: found SDCA function type 1 ADR 4

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: sdca: add initial module

Add new module for SDCA (SoundWire Device Class for Audio) support.
For now just add a mechanism to allocate a context.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: add SDCA alloc/parsing

Before the driver probe, we can parse the SDCA information in platform
firmware to discover function-related information.

The SDCA information is abstracted away since:
a) not all implementations require SDCA support.
b) SDCA support will require tight integration with ASoC, specifically
the DAPM/DAI parts.

The SDCA integration follows the USB model, where the Audio Class is
handled by the sound/ subsystem. The ASoC codec drivers are the
intended users of the SDCA information reported by ACPI.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
v2: all the SDCA information is moved to a sound/soc/sdca library, and the information is parsed before the driver probe (if present).

bardliao commented:
Thanks @plbossart Looks like a good start for SDCA driver.

plbossart commented:
@shumingfan would you happen to know what this Function 4 might be. I wonder if the function topology uses an older version of the SDCA/DisCo spec?

plbossart committed:
ASoC: sdca: parse entities

For now only extract the ID, label and entity_type.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: sdca: extract optional topology_features mask

In contrast to the USB spec, the SDCA only allows for well-defined
connections between entities. The intent was to limit errors/support
with a limited set of configurations. In other words, connections that
are not described in the SDCA spec are not permitted.

In order to allow for product variations, optional paths can be
supported. These options are described with a topology_features mask.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
logs with the last version

````
[   51.528012] rt1316-sdca sdw:0:1:025d:1316:01: found SDCA function AF04 type 1 ADR 4
[   51.528023] rt1316-sdca sdw:0:1:025d:1316:01: AF04: SDCA function has topology-features mask 0xb00
[   51.528033] acpi device:28: find_sdca_entities: AF04: found entity 0x1 type 0x2 label IT 21
[   51.528038] acpi device:28: find_sdca_entities: AF04: found entity 0x2 type 0x21 label UDMPU 21
[   51.528043] acpi device:28: find_sdca_entities: AF04: found entity 0x3 type 0x7 label FU 21
[   51.528047] acpi device:28: find_sdca_entities: AF04: found entity 0x33 type 0x11 label PDE 22
[   51.528052] acpi device:28: find_sdca_entities: AF04: found entity 0x22 type 0xa label XU 22
[   51.528056] acpi device:28: find_sdca_entities: AF04: found entity 0x29 type 0x24 label SAPU 29
[   51.528061] acpi device:28: find_sdca_entities: AF04: found entity 0x54 type 0x21 label UDMPU 23
[   51.528065] acpi device:28: find_sdca_entities: AF04: found entity 0x55 type 0x7 label FU 23
[   51.528070] acpi device:28: find_sdca_entities: AF04: found entity 0x31 type 0x11 label PDE 23
[   51.528074] acpi device:28: find_sdca_entities: AF04: found entity 0x56 type 0x3 label OT 23
[   51.528078] acpi device:28: find_sdca_entities: AF04: found entity 0x51 type 0x2 label IT 29
[   51.528082] acpi device:28: find_sdca_entities: AF04: found entity 0x34 type 0x11 label PDE 24
[   51.528086] acpi device:28: find_sdca_entities: AF04: found entity 0x24 type 0xa label XU 24
[   51.528091] acpi device:28: find_sdca_entities: AF04: found entity 0x52 type 0x3 label OT 24
[   51.528095] acpi device:28: find_sdca_entities: AF04: found entity 0x53 type 0xb label CS 24
[   51.528100] acpi device:28: find_sdca_entities: AF04: found entity 0x21 type 0xb label CS 21
[   51.528104] acpi device:28: find_sdca_entities: AF04: found entity 0x40 type 0x30 label TG 23
[   51.528638] rt1316-sdca sdw:0:2:025d:1316:01: found SDCA function AF04 type 1 ADR 4
[   51.528643] rt1316-sdca sdw:0:2:025d:1316:01: AF04: SDCA function has topology-features mask 0xb00
[   51.528651] acpi device:2a: find_sdca_entities: AF04: found entity 0x1 type 0x2 label IT 21
[   51.528655] acpi device:2a: find_sdca_entities: AF04: found entity 0x2 type 0x21 label UDMPU 21
[   51.528659] acpi device:2a: find_sdca_entities: AF04: found entity 0x3 type 0x7 label FU 21
[   51.528665] acpi device:2a: find_sdca_entities: AF04: found entity 0x33 type 0x11 label PDE 22
[   51.528668] acpi device:2a: find_sdca_entities: AF04: found entity 0x22 type 0xa label XU 22
[   51.528673] acpi device:2a: find_sdca_entities: AF04: found entity 0x29 type 0x24 label SAPU 29
[   51.528677] acpi device:2a: find_sdca_entities: AF04: found entity 0x54 type 0x21 label UDMPU 23
[   51.528681] acpi device:2a: find_sdca_entities: AF04: found entity 0x55 type 0x7 label FU 23
[   51.528685] acpi device:2a: find_sdca_entities: AF04: found entity 0x31 type 0x11 label PDE 23
[   51.528689] acpi device:2a: find_sdca_entities: AF04: found entity 0x56 type 0x3 label OT 23
[   51.528693] acpi device:2a: find_sdca_entities: AF04: found entity 0x51 type 0x2 label IT 29
[   51.528698] acpi device:2a: find_sdca_entities: AF04: found entity 0x34 type 0x11 label PDE 24
[   51.528712] acpi device:2a: find_sdca_entities: AF04: found entity 0x24 type 0xa label XU 24
[   51.528718] acpi device:2a: find_sdca_entities: AF04: found entity 0x52 type 0x3 label OT 24
[   51.528722] acpi device:2a: find_sdca_entities: AF04: found entity 0x53 type 0xb label CS 24
[   51.528726] acpi device:2a: find_sdca_entities: AF04: found entity 0x21 type 0xb label CS 21
[   51.528730] acpi device:2a: find_sdca_entities: AF04: found entity 0x40 type 0x30 label TG 23
[   51.529357] rt712-sdca sdw:0:0:025d:0713:01: found invalid SDCA function AF01 type 4 ADR 1, skipped
[   51.529374] rt712-sdca sdw:0:0:025d:0713:01: found SDCA function AF03 type 6 ADR 3
[   51.529383] acpi device:26: find_sdca_entities: AF03: found entity 0x1 type 0x31 label (null)
````

shumingfan commented:
> @shumingfan would you happen to know what this Function 4 might be. I wonder if the function topology uses an older version of the SDCA/DisCo spec?

I will try to ask the hardware team about this. If there are any results, I will give feedback.

plbossart committed:
ASoC: sdca: read 'function-busy-max-delay' property and report error if non-zero

In their infinite wisdom, the MIPI SDCA subgroup allowed the semantics
of the "Command Ignored" to be overloaded.

In the MIPI SoundWire 1.x spec, the 'Command Ignored' can only happen
in three cases:
a) as a result of an access to an invalid register
b) as a result of an access to a device that is not attached to the bus
c) when enumeration is complete.

For SDCA devices, the 'Command Ignored' might be a valid response if
the device is currently 'busy' dealing with the previous
transaction. In that case, software has to poll the 'Function_Busy'
bit and re-issue the read/write command when that bit is no longer
asserted.

This behavior is currently NOT supported in the regmap and default
read/write functions, and there is currently no hardware which
implements this feature. For now the only thing we can do is to log an
error and ask end-users to report this problematic scenario.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: sdca: parse entity input pins

The DisCo specification describes the connections between entities
with a 'pin-input-list' mask, and subproperties to point to the
connected entity. Unfortunately there's no direct access to the
connected entity id, so we have to search the entity list using the
'label' string, which is the only piece of information useable for a
match.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

vijendarmukunda commented:
Thanks @plbossart for pushing the SDCA driver patches. We will also try to validate on AMD platform.

plbossart committed:
ASoC: sdca: extract function initialization table

The MIPI DisCo 2.1 spec defines an "initialization table", which
contains a list of 4-byte addresses and 1 byte value. The registers
need to be updated during the initial boot and if context is lost.

Note that the data in the table is represented as an ACPI "buffer",
but the buffer is accessible with a DSD property:

ToUUID ("edb12dd0-363d-4085-a3d2-49522ca160c4")
Package (0x01)
{
  Package (0x02)
  {
    "mipi-sdca-function-initialization-table",
    "BUF0"
  }
}

Name (BUF0, Buffer (0x01BD)
{
  0x20, 0xC7, 0x00, 0x00, 0x17, 0x21, 0xC7, 0x00,
  0x00, 0x00, 0x22, 0xC7, 0x00, 0x00, 0x3E, 0x23,
  0xC7, 0x00, 0x00, 0x06, 0x24, 0xC7, 0x00, 0x00,
...
}

The connection between _DSD and ACPI is defined in the 'Data Buffer'
extension [1] already supported in the Linux kernel.

[1] https://github.com/UEFI/DSD-Guide/blob/main/src/dsd-guide.adoc#buffer-data-extension-uuid

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

shumingfan commented:
> I will try to ask the hardware team about this. If there are any results, I will give feedback.

The hardware engineer thinks reporting the function type 4 is a mistake.
The rt713 only needs the function type 6 actually. The rt1713 should report the mic function.

plbossart commented:
> The hardware engineer thinks reporting the function type 4 is a mistake. The rt713 only needs the function type 6 actually. The rt1713 should report the mic function.

Right, that's what I thought @shumingfan. In this device there's no RT1713 (local mic disabled), so I can't check further.

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -30,5 +30,85 @@ int sdca_alloc(void **context,
+	u64 addr;
+
+	status = acpi_evaluate_integer(adev->handle,
+				       METHOD_NAME__ADR, NULL, &addr);
```
Should probably acpi_get_local_address

plbossart replied:
Erm, no. that function is wrong. _ADR is a 64-bit value, but the code assumes it's 32-bits.

I vaguely remember we had to fix some of the ACPI subsystem to make _ADR a true 64 bit value, this function has not been updated apparently.

Actually now I do remember, and it's even in git:

````
commit ca6f998cf9a259fe6019d44768064e0cfe8a925b
Author: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
Date:   Wed May 1 07:53:22 2019 -0500

    ACPI: bus: change _ADR representation to 64 bits
````

So this code would need to be updated to u64.

````c
int acpi_get_local_address(acpi_handle handle, u32 *addr)
{
	unsigned long long adr;
	acpi_status status;

	status = acpi_evaluate_integer(handle, METHOD_NAME__ADR, NULL, &adr);
	if (ACPI_FAILURE(status))
		return -ENODATA;

	*addr = (u32)adr;
	return 0;
}
````

Gah.

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -191,6 +192,18 @@ static int find_sdca_function(struct acpi_device *adev, void *data)
+			"%pfwP: unsupported non-zero delay %d us for Function_Busy\n",
+			function_node, max_busy_delay);
+		dev_err(parent,
+			"please report this on https://github.com/thesofproject/linux/issues\n");
```
Well you say there is no hardware that uses this feature, but cs42l43 and likely its successors make heavy use of it if they are used in SDCA mode. Not that support needs to be added right now, but just making you aware.

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -119,6 +119,140 @@ static int find_sdca_entities(struct device *dev,
+		return -EINVAL;
+	}
+
+	for(pin = 1; pin < 64; pin++) {
```
Probably slightly nice to use BITS_PER_TYPE(input_pin_list)

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -119,6 +119,140 @@ static int find_sdca_entities(struct device *dev,
+		struct fwnode_handle *connected_node;
+		struct sdca_entity *connected_entity;
+		const char *connected_label;
+		char pin_property[40];
```
I would be tempted to add some sort of define for this max length, lot of max 40 byte strings used in the code.

plbossart commented:
closing for now, will re-add the functionality when basic parsing is implemented with PR #4995
